File: fl-server/project_fl/privacy.py
"""
Differential privacy utilities.

Adds calibrated Gaussian noise to model weights before aggregation so that
the server cannot reconstruct individual vehicle data from the updates.
Privacy budget: ε ≈ 0.5 (tight), δ = 1e-5.
"""
import random
import math
from typing import List


def _gaussian_noise(sigma: float) -> float:
    """Box-Muller transform for zero-mean Gaussian noise."""
    u1 = random.random() or 1e-10
    u2 = random.random() or 1e-10
    return sigma * math.sqrt(-2.0 * math.log(u1)) * math.cos(2.0 * math.pi * u2)


def add_gaussian_noise(layer: List[float], sigma: float = 0.5) -> List[float]:
    """
    Add Gaussian noise N(0, sigma²) to every parameter in a weight layer.
    Called server-side on each received update before storage.
    """
    return [w + _gaussian_noise(sigma) for w in layer]


def clip_weights(layer: List[float], clip_norm: float = 1.0) -> List[float]:
    """
    Clip weight layer to L2 norm ≤ clip_norm (sensitivity bounding).
    Applied before noise addition for tighter privacy guarantees.
    """
    norm = math.sqrt(sum(w * w for w in layer)) or 1e-10
    if norm <= clip_norm:
        return layer
    scale = clip_norm / norm
    return [w * scale for w in layer]


def privatise_weights(
    layer: List[float],
    clip_norm: float = 1.0,
    sigma: float = 0.5,
) -> List[float]:
    """Full DP pipeline: clip → add noise."""
    clipped = clip_weights(layer, clip_norm)
    return add_gaussian_noise(clipped, sigma)

File: mobile-app/utils/privacy.ts
/**
 * Client-side privacy utilities.
 *
 * Before model weights leave the device we:
 *  1. Clip each layer's L2 norm (sensitivity bounding)
 *  2. Add Gaussian noise  N(0, σ²)   (local differential privacy)
 *
 * This is defence-in-depth on top of the server-side noise in privacy.py.
 */

// ── Gaussian noise (Box-Muller) ───────────────────────────────────────────────

function gaussianNoise(sigma: number): number {
  const u1 = Math.random() || 1e-10;
  const u2 = Math.random() || 1e-10;
  return sigma * Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2);
}

// ── L2 norm ───────────────────────────────────────────────────────────────────

function l2Norm(layer: number[]): number {
  return Math.sqrt(layer.reduce((acc, w) => acc + w * w, 0)) || 1e-10;
}

// ── Public API ────────────────────────────────────────────────────────────────

/**
 * Clip a weight layer so its L2 norm ≤ clipNorm.
 * Prevents any single update from dominating aggregation.
 */
export function clipLayer(layer: number[], clipNorm = 1.0): number[] {
  const norm = l2Norm(layer);
  if (norm <= clipNorm) return layer;
  const scale = clipNorm / norm;
  return layer.map(w => w * scale);
}

/**
 * Add calibrated Gaussian noise to a weight layer.
 * σ = 0.3 by default (lighter than server-side 0.5 to preserve utility).
 */
export function addNoise(layer: number[], sigma = 0.3): number[] {
  return layer.map(w => w + gaussianNoise(sigma));
}

/**
 * Full client-side DP pipeline applied to all weight layers before upload.
 */
export function privatiseWeights(
  weights: number[][],
  clipNorm = 1.0,
  sigma = 0.3
): number[][] {
  return weights.map(layer => addNoise(clipLayer(layer, clipNorm), sigma));
}

/**
 * Verify no raw feature data is present in a payload about to be sent.
 * Returns true if safe, throws if raw data detected.
 */
export function assertNoRawData(payload: Record<string, unknown>): true {
  const forbidden = ['soc_start', 'soc_end', 'temperature_c', 'distance_km', 'user_id_raw'];
  for (const key of forbidden) {
    if (key in payload) {
      throw new Error(`Privacy violation: raw field "${key}" found in upload payload`);
    }
  }
  return true;
}
